Update TypeScript to v4 and fix noUncheckedIndexedAccess

Hello @mrbar42 

I personally prefer this library and the API then the others. It has been a while, let's make some update.

TypeScript v4+ introduce some useful additional checks. [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess) is one of them. Since TS import original source `index.ts` instead of `index.d.ts + index.js` for this library, the check will produce error when we import this library.

<details>
  <summary>Error details here</summary>

```log
node_modules/node-graceful/index.ts:142:49 - error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '((event: any) => void) | undefined' is not assignable to parameter of type '(...args: any[]) => void'.
      Type 'undefined' is not assignable to type '(...args: any[]) => void'.

142                 process.on(deadlySignal as any, Graceful.signalsListeners[deadlySignal]);
                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  node_modules/@types/node/process.d.ts:399:17
    399                 on(event: string | symbol, listener: (...args: any[]) => void): this;
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    The last overload is declared here.

node_modules/node-graceful/index.ts:148:65 - error TS2345: Argument of type '((event: any) => void) | undefined' is not assignable to parameter of type '(...args: any[]) => void'.
  Type 'undefined' is not assignable to type '(...args: any[]) => void'.

148                     process.removeListener(deadlySignal as any, Graceful.signalsListeners[deadlySignal]);
                                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors.
```

</details>

I'll send a PR and I think it still working fine with TS v3+
Later on, if you are ok, I can update the depreciated dependencies and try on new NodeJS.

--- a/index.js
+++ b/index.js
@@ -22,7 +22,7 @@ var Graceful = /** @class */ (function () {
                 process.removeListener('uncaughtException', Graceful.exceptionListener);
             }
         },
-        enumerable: true,
+        enumerable: false,
         configurable: true
     });
     Object.defineProperty(Graceful, "captureRejections", {
@@ -40,7 +40,7 @@ var Graceful = /** @class */ (function () {
                 process.removeListener('unhandledRejection', Graceful.rejectionListener);
             }
         },
-        enumerable: true,
+        enumerable: false,
         configurable: true
     });
     Graceful.on = function (signal, listener) {
@@ -116,8 +116,9 @@ var Graceful = /** @class */ (function () {
     Graceful.updateRegistration = function () {
         if (Graceful.listeners.length && !Graceful.isRegistered) {
             var _loop_1 = function (deadlySignal) {
-                Graceful.signalsListeners[deadlySignal] = function () { return Graceful.onDeadlyEvent(deadlySignal); };
-                process.on(deadlySignal, Graceful.signalsListeners[deadlySignal]);
+                var listener = function () { return Graceful.onDeadlyEvent(deadlySignal); };
+                Graceful.signalsListeners[deadlySignal] = listener;
+                process.on(deadlySignal, listener);
             };
             for (var _i = 0, _a = Graceful.DEADLY_SIGNALS; _i < _a.length; _i++) {
                 var deadlySignal = _a[_i];
@@ -128,8 +129,9 @@ var Graceful = /** @class */ (function () {
         else if (!Graceful.listeners.length && Graceful.isRegistered) {
             for (var _b = 0, _c = Graceful.DEADLY_SIGNALS; _b < _c.length; _b++) {
                 var deadlySignal = _c[_b];
-                if (Graceful.signalsListeners[deadlySignal]) {
-                    process.removeListener(deadlySignal, Graceful.signalsListeners[deadlySignal]);
+                var listener = Graceful.signalsListeners[deadlySignal];
+                if (listener) {
+                    process.removeListener(deadlySignal, listener);
                     delete Graceful.signalsListeners[deadlySignal];
                 }
             }

--- a/index.ts
+++ b/index.ts
@@ -138,14 +138,16 @@ export default class Graceful {
     private static updateRegistration() {
         if (Graceful.listeners.length && !Graceful.isRegistered) {
             for (const deadlySignal of Graceful.DEADLY_SIGNALS) {
-                Graceful.signalsListeners[deadlySignal] = () => Graceful.onDeadlyEvent(deadlySignal);
-                process.on(deadlySignal as any, Graceful.signalsListeners[deadlySignal]);
+                const listener = () => Graceful.onDeadlyEvent(deadlySignal);
+                Graceful.signalsListeners[deadlySignal] = listener
+                process.on(deadlySignal as any, listener);
             }
             Graceful.isRegistered = true
         } else if (!Graceful.listeners.length && Graceful.isRegistered) {
             for (const deadlySignal of Graceful.DEADLY_SIGNALS) {
-                if (Graceful.signalsListeners[deadlySignal]) {
-                    process.removeListener(deadlySignal as any, Graceful.signalsListeners[deadlySignal]);
+                const listener = Graceful.signalsListeners[deadlySignal]
+                if (listener) {
+                    process.removeListener(deadlySignal, listener);
                     delete Graceful.signalsListeners[deadlySignal];
                 }
             }

--- a/package.json
+++ b/package.json
@@ -41,8 +41,8 @@
   "license": "MIT",
   "devDependencies": {
     "@types/node": "^12.12.30",
-    "ts-node": "^8.7.0",
-    "tslint": "^6.1.0",
-    "typescript": "^3.8.3"
+    "ts-node": "^10.2.0",
+    "tslint": "^6.1.3",
+    "typescript": "^4.3.5"
   }
 }

--- a/tests/TypeScript.ts
+++ b/tests/TypeScript.ts
@@ -1,10 +1,10 @@
-import Graceful, {GracefulListener} from "../";
+import Graceful from "../";
 
 const listener = (signal: string, event?: object) => void 0;
 const listenerPromise = (signal: string, event?: object) => Promise.resolve("abc");
 const listenerPromise1 = (signal: string, event?: object) => Promise.resolve({});
-const listenerPromise2 = (signal: string, event?: object) => new Promise((resolve) => resolve());
-const listenerPromise3 = (signal: string, event?: object) => new Promise((resolve) => resolve());
+const listenerPromise2 = (signal: string, event?: object) => new Promise<void>((resolve) => resolve());
+const listenerPromise3 = (signal: string, event?: object) => new Promise<void>((resolve) => resolve());
 const listenerPromise4 = () => null;
 
 Graceful.on("exit", listener);

--- a/tests/wait-for-multiple-promise.ts
+++ b/tests/wait-for-multiple-promise.ts
@@ -2,7 +2,7 @@
 
 import Graceful from "../";
 
-const handlerSuccess = () => {
+const handlerSuccess = (): Promise<void> => {
     return new Promise((resolve) =>  {
         setTimeout(() => {
             process.stdout.write("ok");
@@ -11,8 +11,8 @@ const handlerSuccess = () => {
     });
 };
 
-const handlerReject = () => {
-    return new Promise((resolve, reject) => {
+const handlerReject = (): Promise<void> => {
+    return new Promise((_resolve, reject) => {
         setTimeout(() => {
             process.stdout.write("ok");
             reject();

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -5,6 +5,8 @@
     "strict": true,
     "alwaysStrict": true,
     "preserveConstEnums": true,
+    "noPropertyAccessFromIndexSignature": true,
+    "noUncheckedIndexedAccess": true,
     "sourceMap": false,
     "moduleResolution": "Node",
     "esModuleInterop": true,

